components/style/values/specified/position.rs

```diff
@@ -7,278 +7,294 @@
-
 /// The specified value of a CSS `<position>`
-pub type Position = PositionWithKeyword<PositionValue<LengthOrPercentage>>;
+pub type Position = GenericPosition<HorizontalPosition, VerticalPosition>;
+
+/// The specified value of a horizontal position.
+pub type HorizontalPosition = PositionComponent<X>;
+
+/// The specified value of a vertical position.
+pub type VerticalPosition = PositionComponent<Y>;
+
+/// The specified value of a component of a CSS `<position>`.
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[derive(Clone, Debug, PartialEq)]
+pub enum PositionComponent<S> {
+    /// `center`
+    Center,
+    /// `<lop>`
+    Length(LengthOrPercentage),
+    /// `<side> <lop>?`
+    Side(S, Option<LengthOrPercentage>),
+}
 
-/// The specified value for `<position>` values without a keyword.
-pub type OriginPosition = GenericPosition<LengthOrPercentage, LengthOrPercentage>;
+define_css_keyword_enum! { X:
+    "left" => Left,
+    "right" => Right,
+}
+add_impls_for_keyword_enum!(X);
+
+define_css_keyword_enum! { Y:
+    "top" => Top,
+    "bottom" => Bottom,
+}
+add_impls_for_keyword_enum!(Y);
 
-impl Parse for OriginPosition {
+impl Parse for Position {
     fn parse(context: &ParserContext, input: &mut Parser) -> Result<Self, ()> {
-        let result = parse_origin(context, input)?;
-        match result.depth {
-            Some(_) => Err(()),
-            None => Ok(GenericPosition {
-                horizontal: result.horizontal.unwrap_or(LengthOrPercentage::Percentage(Percentage(0.5))),
-                vertical: result.vertical.unwrap_or(LengthOrPercentage::Percentage(Percentage(0.5))),
-            })
-        }
+        Self::parse_quirky(context, input, AllowQuirks::No)
     }
 }
 
-type PositionComponent = Either<LengthOrPercentage, Keyword>;
-
 impl Position {
-    /// Create a new position value from either a length or a keyword.
-    pub fn from_components(mut first_position: Option<PositionComponent>,
-                           mut second_position: Option<PositionComponent>,
-                           first_keyword: Option<PositionComponent>,
-                           second_keyword: Option<PositionComponent>) -> Result<Position, ()> {
-        // Unwrap for checking if values are at right place.
-        let first_key = first_keyword.clone().unwrap_or(Either::Second(Keyword::Left));
-        let second_key = second_keyword.clone().unwrap_or(Either::Second(Keyword::Top));
-
-        let (horiz_keyword, vert_keyword) = match (&first_key, &second_key) {
-            // Check if a position is specified after center keyword.
-            (&Either::Second(Keyword::Center), _) if first_position.is_some() => return Err(()),
-            (_, &Either::Second(Keyword::Center)) if second_position.is_some() => return Err(()),
-
-            // Check first and second keywords for both 2 and 4 value positions.
-
-            // FIXME(canaltinova): Allow logical keywords for Position. They are not in current spec yet.
```

Comment:
They shouldn't be supported here, if we ever support them, this will be only in `background-position{,-x,-y}`, `<position>` as specified in [css-values-4](https://drafts.csswg.org/css-values-4/#typedef-position) doesn't have logical keywords.